metagraph/src/graph/representation/succinct/boss.cpp

```diff
@@ -1486,43 +1484,40 @@ uint64_t traverse_dummy_edges(const BOSS &graph,
                               sdsl::bit_vector *redundant_mask,
                               sdsl::bit_vector *traversed_mask,
                               size_t num_threads,
-                              bool verbose) {
+                              const std::function<void(edge_index, size_t)> &on_dummy
+                                        = [](edge_index, size_t) {}) {
     assert(!redundant_mask || redundant_mask->size() == graph.get_W().size());
     assert(!traversed_mask || traversed_mask->size() == graph.get_W().size());
 
+    // assume the main dummy source node already traversed
+    std::atomic<uint64_t> num_dummy_traversed { 1 };
+    on_dummy(1, 0); // call the main dummy source edge
     if (traversed_mask)
         (*traversed_mask)[1] = true;
 
     if (graph.get_last(1))
```

Comment:
maybe add a short comment, as I kept wondering how this thing works for cycles until I saw this if:
`if (graph.get_last(1)) // graph has no dummy k-mers`

Reply:
will do